Mathlib/Analysis/InnerProductSpace/Harmonic/Basic.lean

```diff
@@ -0,0 +1,157 @@
+    ∀ᶠ y in 𝓝 x, HarmonicAt f y := by
+  filter_upwards [h.1.eventually (by aesop), h.2.eventually_nhds] with a h₁a h₂a
+  exact ⟨h₁a, h₂a⟩
+
+variable (f) in
+/--
+Harmonicity is an open property.
+-/
+theorem harmonicAt_isOpen : IsOpen { x : E | HarmonicAt f x } :=
+  isOpen_iff_mem_nhds.2 (fun _ hx ↦ hx.eventually)
+
+/--
+If `f` is harmonic in a neighborhood of `s`, it is harmonic in a neighborhood of
+every subset.
+-/
+lemma HarmonicOnNhd.mono (h : HarmonicOnNhd f s) (hst : t ⊆ s) :
+    HarmonicOnNhd f t := fun x hx ↦ h x (hst hx)
+
+/-!
+## Vector Space Structure
+-/
+
+/--
+Sums of harmonic functions are harmonic.
+-/
+theorem HarmonicAt.add (h₁ : HarmonicAt f₁ x) (h₂ : HarmonicAt f₂ x) :
+    HarmonicAt (f₁ + f₂) x := by
+  constructor
+  · exact h₁.1.add h₂.1
+  · filter_upwards [h₁.1.laplacian_add_nhds h₂.1, h₁.2, h₂.2] with a h₁a h₂a h₃a
+    simp_all
+
+/--
+Sums of harmonic functions are harmonic.
+-/
+theorem HarmonicOnNhd.add (h₁ : HarmonicOnNhd f₁ s) (h₂ : HarmonicOnNhd f₂ s) :
+    HarmonicOnNhd (f₁ + f₂) s := fun x hx ↦ (h₁ x hx).add (h₂ x hx)
+
+/--
+Scalar multiples of harmonic functions are harmonic.
+-/
+theorem HarmonicAt.const_smul (h : HarmonicAt f x) :
+    HarmonicAt (c • f) x := by
+  constructor
+  · exact h.1.const_smul c
+  · filter_upwards [laplacian_smul_nhds c h.1, h.2] with a h₁a h₂a
+    simp_all
+
+/--
+Scalar multiples of harmonic functions are harmonic.
+-/
+theorem HarmonicOnNhd.const_smul (h : HarmonicOnNhd f s) :
+    HarmonicOnNhd (c • f) s := fun x hx ↦ (h x hx).const_smul
+
+/-!
+## Compatibility with Linear Maps
+-/
+
+/--
+Compositions of continuous `ℝ`-linear maps with harmonic functions are harmonic.
+-/
+theorem HarmonicAt.comp_CLM (h : HarmonicAt f x) (l : F →L[ℝ] G) :
+    HarmonicAt (l ∘ f) x := by
+  constructor
+  · exact h.1.continuousLinearMap_comp l
+  · filter_upwards [h.1.laplacian_CLM_comp_left_nhds (l := l), h.2] with a h₁a h₂a
+    simp_all
+
+/--
+Compositions of continuous linear maps with harmonic functions are harmonic.
+-/
+theorem HarmonicOnNhd.comp_CLM (h : HarmonicOnNhd f s) (l : F →L[ℝ] G) :
+    HarmonicOnNhd (l ∘ f) s := fun x hx ↦ (h x hx).comp_CLM l
+
+/--
+Functions are harmonic iff their compositions with continuous linear
+equivalences are harmonic.
+-/
+theorem harmonicAt_iff_harmonicAt_comp_CLE (l : F ≃L[ℝ] G) :
+    HarmonicAt f x ↔ HarmonicAt (l ∘ f) x := by
+  constructor
+  · exact fun h ↦ h.comp_CLM l.toContinuousLinearMap
+  · have : f = l.symm.toContinuousLinearMap ∘ l ∘ f := by
+      unfold Function.comp
+      simp
+    nth_rewrite 2 [this]
+    exact fun h ↦ h.comp_CLM l.symm.toContinuousLinearMap
+
+/--
+Functions are harmonic iff their compositions with continuous linear
+equivalences are harmonic.
+-/
+theorem harmonicOnNhd_iff_harmonicOnNhd_comp_CLE (l : F ≃L[ℝ] G) :
+    HarmonicOnNhd f s ↔ HarmonicOnNhd (l ∘ f) s := by
+  constructor
+  <;> exact fun h x hx ↦ by simpa [← harmonicAt_iff_harmonicAt_comp_CLE] using h x hx
```

Comment:
Usually we would put `<;>` at the end of the preceding line. This can be golfed down to one line anyway:
```suggestion
  forall₂_congr fun _ _ ↦ harmonicAt_iff_harmonicAt_comp_CLE l
```